tests/testbed_setup/test_gen_spy_testbed.py

```diff
@@ -0,0 +1,109 @@
+import logging
+import json
+import os
+import pytest
+import yaml
+
+from collections import defaultdict
+from jinja2 import Template
+
+
+TESTBED_TEMPLATE = "templates/spytest_testbed.yaml.j2"
+PTF_INTERFACE_TEMPLATE = "1/%d"
+
+pytestmark = [
+    pytest.mark.topology("util"),
+    pytest.mark.sanity_check(skip_sanity=True)
+]
+
+
+@pytest.fixture(scope="function")
+def hostvars(duthosts):
+    """Return host variables dicts for DUTs defined in testbed."""
+    if not duthosts:
+        return {}
+    var_manager = duthosts[-1].host.options["variable_manager"]
+    hostvars_all = var_manager.get_vars()["hostvars"]
+    return {duthost.hostname: hostvars_all[duthost.hostname]
+            for duthost in duthosts}
+
+
+def test_gen_spy_testbed(conn_graph_facts_multi_duts, hostvars, testbed,
+                         pytestconfig):
+    """Generate spytest testbed file."""
+
+    def _interface_key(interface):
+        """Get interface key to sort."""
+        return list(map(int, interface.lstrip("Ethernet").split("/")))
+
+    def _to_string(obj):
+        """Convert unicodes in obj to strings"""
+        return yaml.safe_load(json.dumps(obj))
+
+    hostnames = testbed["duts"]
+    device_conn = _to_string(conn_graph_facts_multi_duts["device_conn"])
+    connections = dict(
+        zip(hostnames, conn_graph_facts_multi_duts["device_conn"]))
+
+    # devices section
+    devices = []
+    for hostname in hostnames:
+        hostvar = hostvars[hostname]
+        login_info = {}
+        login_info["login_access"] = \
+            json.dumps(hostvar["login_access"]).replace('"', '')
+        login_info["login_credentials"] = \
+            json.dumps(hostvar["login_credentials"]).replace('"', '')
+        devices.append((hostname, login_info))
+
+    # topology section
+    ptf_connections = []
+    intf = 1
+    for hostname in hostnames:
+        end_device = hostname
+        conns = connections[hostname]
+        end_ports = sorted(
+            (_ for _ in conns if conns[_]['peerdevice'] not in connections),
+            key=_interface_key)
+        for end_port in end_ports:
+            ptf_conn = {
+                "start_port": PTF_INTERFACE_TEMPLATE % intf,
```

Comment:
This requires the DUTs to be connected to the PTF sequentially, right? Do we have that guarantee in the testbed file?

Reply:
Yes, the interfaces in PTF are inserted sequentially in the same order as the source DUTs defined in `testbed.csv`.
For instance, the `testbed.csv` contains an entry like:
```
spytest,spytest,fullmesh,docker-ptf-brcm,ptf-unknown,192.168.0.1/24,,server_16,,[lab-msn2700-spy-1;lab-msn2700-spy-2;lab-a7060-spy-1;lab-a7060-spy-2],hellospytest
```
And the `device_vlan_list` for duts returned by `conn_graph_facts` will be like `[[190, 191, 192, 193], [202, 203, 204, 205], [206, 207, 208, 209], [210, 211, 212, 213]]`. The interfaces will be `Eth0` ~ `Ether15`, which are in the same order as flattened `device_vlan_list`.

Reply:
Cool, thank you for the explanation! :)